File: lib/studentMatcher.ts
import Fuse from 'fuse.js'

export interface StudentMatchCandidate {
  id: string
  name: string
  class_name?: string
  className?: string
  school_name?: string
  schoolName?: string
  score?: number
  matricula?: string
  portal_native_id?: string
  rollNumber?: string
}

export type MatchResultStatus = 'exact' | 'confident_match' | 'ambiguous' | 'not_found'

export interface StudentMatchResult {
  status: MatchResultStatus
  student: StudentMatchCandidate | null
  candidates: StudentMatchCandidate[]
  confidence: number
  disambiguationPrompt?: string
}

/**
 * Remove acentos e normaliza para comparação sem ruído fonético
 */
export function normalizeStudentName(name: string): string {
  return (name || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .trim()
}

/**
 * Resolve o aluno correto a partir de um nome digitado ou falado.
 * Implementa matching em 4 vias:
 * 1. Matrícula prioritária (portal_native_id)
 * 2. Nome exato (+ Turma se disponível)
 * 3. Substring / Token Overlap
 * 4. Primeiro Nome / Prefixo
 * NUNCA escolhe silenciosamente quando há ambiguidade.
 */
export function matchStudentByName(
  queryName: string,
  students: StudentMatchCandidate[],
  options?: {
    threshold?: number
    ambiguityDelta?: number
    matricula?: string
    class_ref?: string
  }
): StudentMatchResult {
  const qClean = normalizeStudentName(queryName)
  const cleanMat = String(options?.matricula || '').trim()

  if ((!qClean && !cleanMat) || !students || students.length === 0) {
    return {
      status: 'not_found',
      student: null,
      candidates: [],
      confidence: 0,
      disambiguationPrompt: 'Nenhum aluno encontrado na lista.'
    }
  }

  // 1. VIA 1: Match Direto por Matrícula / portal_native_id (Determinístico)
  if (cleanMat) {
    const matMatches = students.filter(s => {
      const sMat = String(s.matricula || s.portal_native_id || s.rollNumber || s.id || '').trim()
      return sMat && sMat === cleanMat
    })
    if (matMatches.length === 1) {
      return {
        status: 'exact',
        student: matMatches[0],
        candidates: matMatches,
        confidence: 1.0
      }
    }
  }

  // Match se queryName for a própria matrícula
  const directMatMatches = students.filter(s => {
    const sMat = String(s.matricula || s.portal_native_id || s.rollNumber || '').trim()
    return sMat && sMat.toLowerCase() === qClean
  })
  if (directMatMatches.length === 1) {
    return {
      status: 'exact',
      student: directMatMatches[0],
      candidates: directMatMatches,
      confidence: 1.0
    }
  }

  // 2. VIA 2: Verificação de correspondência exata por Nome
  const exactMatches = students.filter(s => normalizeStudentName(s.name) === qClean)
  if (exactMatches.length === 1) {
    return {
      status: 'exact',
      student: exactMatches[0],
      candidates: [exactMatches[0]],
      confidence: 1.0
    }
  }
  if (exactMatches.length > 1) {
    if (options?.class_ref) {
      const cleanCls = normalizeStudentName(options.class_ref)
      const classFiltered = exactMatches.filter(s => {
        const cls = normalizeStudentName(s.class_name || s.className || '')
        return cls.includes(cleanCls)
      })
      if (classFiltered.length === 1) {
        return {
          status: 'exact',
          student: classFiltered[0],
          candidates: classFiltered,
          confidence: 1.0
        }
      }
    }
    // Mesmo nome exato em turmas diferentes ou sem filtro! Exige desambiguação
    const candidates = exactMatches.map(s => ({ ...s, score: 0 }))
    const candidateLabels = candidates.map(s => `"${s.name}" (${s.class_name || s.className || 'Turma não especificada'})`).join(' ou ')
    return {
      status: 'ambiguous',
      student: null,
      candidates,
      confidence: 0.5,
      disambiguationPrompt: `Existem ${exactMatches.length} alunos com o mesmo nome: ${candidateLabels}. Por favor, informe a turma ou a escola para confirmar.`
    }
  }

  // 2. Busca Fuzzy com Fuse.js
  const preparedList = students.map(s => ({
    ...s,
    normalizedName: normalizeStudentName(s.name)
  }))

  const fuse = new Fuse(preparedList, {
    keys: ['normalizedName', 'name'],
    threshold: options?.threshold ?? 0.35, // Score de 0 (perfeito) a 1 (nenhuma correspondência)
    includeScore: true,
    ignoreLocation: true,
    minMatchCharLength: 2
  })

  const results = fuse.search(qClean)

  if (results.length === 0) {
    return {
      status: 'not_found',
      student: null,
      candidates: [],
      confidence: 0,
      disambiguationPrompt: `Não encontrei nenhum aluno com nome similar a "${queryName}". Verifique a ortografia ou cadastre o aluno.`
    }
  }

  const best = results[0]
  const bestScore = best.score ?? 1.0 // 0 é melhor
  const ambiguityDelta = options?.ambiguityDelta ?? 0.18

  // Se houver mais de um resultado, avalia se há empate ou proximidade de score
  if (results.length > 1) {
    const second = results[1]
    const secondScore = second.score ?? 1.0
    const delta = secondScore - bestScore

    // Se o segundo colocado está muito próximo do primeiro (< ambiguityDelta), temos ambiguidade!
    if (delta < ambiguityDelta || (bestScore > 0.20 && secondScore < 0.40)) {
      const topCandidates = results.slice(0, 3).map(r => ({
        id: r.item.id,
        name: r.item.name,
        class_name: r.item.class_name || r.item.className,
        school_name: r.item.school_name || r.item.schoolName,
        score: r.score
      }))

      if (options?.class_ref) {
        const cleanCls = normalizeStudentName(options.class_ref)
        const classFiltered = topCandidates.filter(c => {
          const cls = normalizeStudentName(c.class_name || '')
          return cls.includes(cleanCls)
        })
        if (classFiltered.length === 1) {
          return {
            status: 'exact',
            student: classFiltered[0],
            candidates: classFiltered,
            confidence: 1.0
          }
        }
      }

      const candidateLabels = topCandidates
        .map(c => `"${c.name}"${c.class_name ? ` (${c.class_name})` : ''}`)
        .join(' ou ')

      return {
        status: 'ambiguous',
        student: null,
        candidates: topCandidates,
        confidence: Number((1 - bestScore).toFixed(2)),
        disambiguationPrompt: `Fiquei em dúvida entre alunos com nomes parecidos: ${candidateLabels}. Qual deles você gostaria de selecionar?`
      }
    }
  }

  // Se o score do melhor resultado for suficientemente confiável (< 0.35)
  if (bestScore <= 0.35) {
    const matchedStudent: StudentMatchCandidate = {
      id: best.item.id,
      name: best.item.name,
      class_name: best.item.class_name || best.item.className,
      school_name: best.item.school_name || best.item.schoolName,
      score: best.score
    }

    return {
      status: 'confident_match',
      student: matchedStudent,
      candidates: [matchedStudent],
      confidence: Number((1 - bestScore).toFixed(2))
    }
  }

  // Score muito fraco (> 0.35)
  return {
    status: 'not_found',
    student: null,
    candidates: results.slice(0, 2).map(r => ({ ...r.item, score: r.score })),
    confidence: Number((1 - bestScore).toFixed(2)),
    disambiguationPrompt: `Não tenho certeza sobre "${queryName}". Você se referia a "${best.item.name}"?`
  }
}

File: sidecar/student_matcher.py
"""
sidecar/student_matcher.py — Motor Unificado de Correspondência e Reconciliação em 4 Vias

Espelha e unifica os princípios de lib/rosterReconciler.ts e lib/studentMatcher.ts:
Princípio Inegociável:
1. O Portal Escolar é a fonte primária de verdade.
2. A matrícula (portal_native_id) é a chave primária estável de identidade.
3. Se houver homônimos ou nomes parecidos sem matrícula para desempate,
   o sistema NUNCA escolhe silenciosamente — retorna status='ambiguous' exigindo confirmação humana.
"""

import re
import unicodedata
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple, Union


def normalize_student_name(name: str) -> str:
    """Remove acentos, pontuação e normaliza caixa para comparação fonética limpa."""
    if not name:
        return ""
    nfkd = unicodedata.normalize("NFD", str(name))
    no_accents = "".join([c for c in nfkd if not unicodedata.combining(c)])
    clean = re.sub(r"[^\w\s]", " ", no_accents)
    return re.sub(r"\s+", " ", clean).strip().lower()


@dataclass
class StudentCandidate:
    id: str
    name: str
    matricula: str = ""
    class_name: str = ""
    score: float = 1.0
    details: str = ""


@dataclass
class MatchResult:
    status: str                         # 'exact' | 'confident_match' | 'ambiguous' | 'not_found'
    student: Optional[StudentCandidate] = None
    candidates: List[StudentCandidate] = field(default_factory=list)
    confidence: float = 0.0
    disambiguation_prompt: Optional[str] = None


def match_student_4_ways(
    query_name: str,
    roster: List[Dict[str, Any]],
    query_matricula: Optional[str] = None,
    class_ref: Optional[str] = None,
    ambiguity_delta: float = 0.18
) -> MatchResult:
    """
    Algoritmo unificado de correspondência em 4 vias:
    1. Match Direto por Matrícula (portal_native_id): 100% determinístico e estável.
    2. Match Exato por Nome Completo Normalizado (+ Turma se disponível).
    3. Match por Substring / Token Overlap (com detecção estrita de empates).
    4. Match por Primeiro Nome / Prefixo (com retorno obrigatório de 'ambiguous' se houver > 1).
    """
    clean_query = normalize_student_name(query_name)
    clean_mat = str(query_matricula).strip() if query_matricula else ""

    if not roster or (not clean_query and not clean_mat):
        return MatchResult(
            status="not_found",
            confidence=0.0,
            disambiguation_prompt="Nenhum aluno informado ou lista vazia."
        )

    candidates_pool: List[StudentCandidate] = []
    for idx, item in enumerate(roster):
        raw_name = str(item.get("name") or item.get("nome") or "")
        raw_mat = str(item.get("matricula") or item.get("portal_native_id") or item.get("rollNumber") or item.get("id") or "")
        c_class = str(item.get("classRef") or item.get("class_name") or item.get("turma") or "")
        det = str(item.get("details") or item.get("fullText") or f"{raw_name} ({raw_mat})").strip()

        candidates_pool.append(StudentCandidate(
            id=raw_mat or f"stu_{idx}",
            name=raw_name,
            matricula=raw_mat,
            class_name=c_class,
            details=det
        ))

    # ──────────────────────────────────────────────────────────────────────────
    # VIA 1: Match Direto por Matrícula (portal_native_id)
    # ──────────────────────────────────────────────────────────────────────────
    if clean_mat:
        mat_matches = [c for c in candidates_pool if c.matricula and c.matricula == clean_mat]
        if len(mat_matches) == 1:
            return MatchResult(
                status="exact",
                student=mat_matches[0],
                candidates=mat_matches,
                confidence=1.0
            )

    # Match direto caso query_name seja a própria matrícula
    raw_query_clean = str(query_name or "").strip().lower()
    if raw_query_clean:
        direct_mat = [
            c for c in candidates_pool
            if c.matricula and c.matricula.strip().lower() == raw_query_clean
        ]
        if len(direct_mat) == 1:
            return MatchResult(
                status="exact",
                student=direct_mat[0],
                candidates=direct_mat,
                confidence=1.0
            )

    # ──────────────────────────────────────────────────────────────────────────
    # VIA 2: Match Exato por Nome Normalizado (+ Turma se informada)
    # ──────────────────────────────────────────────────────────────────────────
    exact_name_matches = [
        c for c in candidates_pool
        if normalize_student_name(c.name) == clean_query
    ]

    if class_ref and len(exact_name_matches) > 1:
        clean_cls = normalize_student_name(class_ref)
        filtered_by_class = [c for c in exact_name_matches if clean_cls in normalize_student_name(c.class_name)]
        if len(filtered_by_class) == 1:
            return MatchResult(
                status="exact",
                student=filtered_by_class[0],
                candidates=filtered_by_class,
                confidence=1.0
            )

    if len(exact_name_matches) == 1:
        return MatchResult(
            status="exact",
            student=exact_name_matches[0],
            candidates=exact_name_matches,
            confidence=1.0
        )
    elif len(exact_name_matches) > 1:
        cand_labels = [f"'{c.name}'" + (f" (#{c.matricula})" if c.matricula else "") for c in exact_name_matches]
        return MatchResult(
            status="ambiguous",
            student=None,
            candidates=exact_name_matches,
            confidence=0.5,
            disambiguation_prompt=f"Existem múltiplos alunos com o mesmo nome exato: {' ou '.join(cand_labels)}. Informe a matrícula para confirmar."
        )

    # ──────────────────────────────────────────────────────────────────────────
    # VIA 3: Substring / Token Overlap
    # ──────────────────────────────────────────────────────────────────────────
    substring_matches = []
    for c in candidates_pool:
        c_norm = normalize_student_name(c.name)
        if clean_query in c_norm or c_norm in clean_query:
            substring_matches.append(c)

    if class_ref and len(substring_matches) > 1:
        clean_cls = normalize_student_name(class_ref)
        filtered_by_class = [c for c in substring_matches if clean_cls in normalize_student_name(c.class_name)]
        if len(filtered_by_class) == 1:
            return MatchResult(
                status="exact",
                student=filtered_by_class[0],
                candidates=filtered_by_class,
                confidence=1.0
            )

    if len(substring_matches) == 1:
        return MatchResult(
            status="confident_match",
            student=substring_matches[0],
            candidates=substring_matches,
            confidence=0.90
        )
    elif len(substring_matches) > 1:
        cand_labels = [f"'{c.name}'" + (f" (#{c.matricula})" if c.matricula else "") for c in substring_matches]
        return MatchResult(
            status="ambiguous",
            student=None,
            candidates=substring_matches,
            confidence=0.5,
            disambiguation_prompt=f"Encontrei mais de um aluno correspondente: {' ou '.join(cand_labels)}. Qual deles você deseja selecionar?"
        )

    # ──────────────────────────────────────────────────────────────────────────
    # VIA 4: Primeiro Nome / Prefixo
    # ──────────────────────────────────────────────────────────────────────────
    first_token = clean_query.split()[0] if clean_query else ""
    if len(first_token) >= 2:
        first_token_matches = [
            c for c in candidates_pool
            if normalize_student_name(c.name).split()[0] == first_token
        ]
        if len(first_token_matches) == 1:
            return MatchResult(
                status="confident_match",
                student=first_token_matches[0],
                candidates=first_token_matches,
                confidence=0.85
            )
        elif len(first_token_matches) > 1:
            cand_labels = [f"'{c.name}'" + (f" (#{c.matricula})" if c.matricula else "") for c in first_token_matches]
            return MatchResult(
                status="ambiguous",
                student=None,
                candidates=first_token_matches,
                confidence=0.5,
                disambiguation_prompt=f"Encontrei {len(first_token_matches)} alunos com o primeiro nome '{first_token.capitalize()}': {' ou '.join(cand_labels)}. Por favor, informe o nome completo ou a matrícula."
            )

    return MatchResult(
        status="not_found",
        student=None,
        candidates=[],
        confidence=0.0,
        disambiguation_prompt=f"Não encontrei nenhum aluno com nome similar a '{query_name}' nesta turma."
    )
